File: java/leetcode/023.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> q = new LinkedList<>();
        
        for(ListNode node: lists) {
            q.offer(node);
        }
        while(q.size() > 1) {
            ListNode tmp1 = q.poll();
            ListNode tmp2 = q.poll();
            q.offer(merge(tmp1, tmp2));
        }
        return q.poll();
    }
    public ListNode merge(ListNode l1, ListNode l2) {
        if(l1 == null) return l2;
        if(l2 == null) return l1;
        
        if(l1.val > l2.val) return merge(l2, l1);
        ListNode head = l1;
        l1.next = merge(l1.next, l2);
        return head;
    }
}

public class Solution {
    public ListNode mergeKLists(List<ListNode> lists) {
        if (lists==null||lists.size()==0) return null;

        PriorityQueue<ListNode> queue= new PriorityQueue<ListNode>(
            lists.size(),new Comparator<ListNode>(){
                @Override
                public int compare(ListNode o1,ListNode o2){
                    if (o1.val<o2.val)
                        return -1;
                    else if (o1.val==o2.val)
                        return 0;
                    else
                        return 1;
                }
            } 
        );

        ListNode dummy = new ListNode(0);
        ListNode tail=dummy;

        for (ListNode node:lists)
            if (node!=null)
                queue.add(node);

        while (!queue.isEmpty()){
            tail.next=queue.poll();
            tail=tail.next;

            if (tail.next!=null)
                queue.add(tail.next);
        }
        return dummy.next;
    }
}

class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        return partition(lists, 0, lists.length - 1);
    }
    public ListNode partition(ListNode[] lists, int s, int e) {
        if(s == e) return lists[s];
        if (s < e) {
            int mid = (e + s) / 2;
            ListNode l1 = partition(lists, s, mid);
            ListNode l2 = partition(lists, mid+1, e);
            return merge(l1, l2);
        } else {
            return null;
        }
    }
    public ListNode merge(ListNode l1, ListNode l2) {
        if (l1 == null) return l2;
        if (l2 == null) return l1;
        if (l1.val < l2.val){
            l1.next = merge(l1.next, l2);
            return l1;
        } else {
            l2.next = merge(l1, l2.next);
            return l2;
        }
    }
}

File: cpp/leetcode/023.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    static bool cmp(ListNode* a, ListNode* b) {
        return a->val < b->val;
    }

    // brute force
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        vector<ListNode*> vec = vector<ListNode*>();
        for (auto it: lists) {
            ListNode* head = it;
            while (head != nullptr) {
                vec.emplace_back(head);
                head = head->next;
            }
        }
        if (vec.size() <= 0) return nullptr;
        sort(vec.begin(), vec.end(), cmp);
        ListNode* head = vec[0];
        ListNode* cur = head;
        for (int i = 1; i < vec.size(); ++i) {
            cur->next = vec[i];
            cur = cur->next;
        }
        cur->next = nullptr;
        return head;
    }


    // priority_queue
    struct cmp_pointer {
        bool operator() (ListNode* a, ListNode* b) {
            return a->val > b->val;
        }
    };
    
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, cmp_pointer> pq;
        for (auto it: lists) {
            ListNode * ptr = it;
            while (ptr != nullptr) {
                pq.push(ptr);
                ptr = ptr->next;
            }
        }
        if (pq.size() <= 0) return nullptr;
        ListNode* head = pq.top(); pq.pop();
        ListNode* cur = head;
        while (pq.size() > 0) {
            ListNode* _cur = pq.top(); pq.pop();
            cur->next = _cur;
            cur = cur->next;
        }
        cur->next = nullptr;
        return head;
    }
};
